### PR TITLE
community: add support for callable filters in FAISS

### PR DESCRIPTION
**Pull request by thiswillbeyourgithub**
- **Description:**
    Filtering in a FAISS vectorstores is very inflexible and doesn't allow that many use case. I think supporting callable like this enables a lot: regular expressions, condition on multiple keys etc. **Note** I had to manually alter a test. I don't understand if it was falty to begin with or if there is something funky going on.
- **Issue:** None
- **Dependencies:** None
- **Twitter handle:** None

### CODE REVIEW

**Review by hwchase17 — APPROVED** [review]
nice! i like this

File: libs/community/tests/unit_tests/vectorstores/test_faiss.py [hunk 1]
```diff
@@ -420,7 +437,11 @@ def test_faiss_with_metadatas_and_filter() -> None:
     )
     assert docsearch.docstore.__dict__ == expected_docstore.__dict__
     output = docsearch.similarity_search("foo", k=1, filter={"page": 1})
-    assert output == [Document(page_content="bar", metadata={"page": 1})]
```
[the remark this answers is not in the document]
Comment (thiswillbeyourgithub) [reply libs/community/tests/unit_tests/vectorstores/test_faiss.py:424]: @hwchase17 I know you just merged it but can you double check if I was right to remove this test?

**Comment by ablacklama** [comment]
@thiswillbeyourgithub @hwchase17 
Just ran into an issue with this. [faiss.py](https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/vectorstores/faiss.py#L333)

on line 333
```python
if filter is not None and filter_func(doc.metadata):
    docs.append((doc, scores[0][j]))
else:
    docs.append((doc, scores[0][j]))
```
It checks whether you have a filtering func and if it's "true", and if you don't have a func, or it's not "true", it returns the result anyways. This means all filtering is completely bypassed. 

I'd recommend changing that code to something like this.
```python
if filter is not None:
    if filter_func(doc.metadata):
        docs.append((doc, scores[0][j]))
else:
    docs.append((doc, scores[0][j]))
```

**Comment by thiswillbeyourgithub** [comment]
I'm on it in a couple hours (I'm so sorry.)

**Comment by ablacklama** [comment]
No worries. I submitted a fix here. https://github.com/langchain-ai/langchain/pull/16967

**Pull request merged**